### ISSUE TITLE
Bogus warning -Wunused-but-set-variable

### ISSUE THREAD

**Issue by JohelEGP** [llvm/llvm-project#54595]
See https://godbolt.org/z/P5Kq4osnd.
```C++
#include <limits>
#include <type_traits>
template<auto F> concept constant_invocable = requires { typename std::integral_constant<int, (F(), 0)>; };
constexpr std::numeric_limits<int> int_limits;
static_assert(!constant_invocable<[] { [i = int_limits.min()]() mutable { (void)--i; }(); }>);
```
```
<source>:5:41: warning: variable 'i' set but not used [-Wunused-but-set-variable]
static_assert(!constant_invocable<[] { [i = int_limits.min()]() mutable { (void)--i; }(); }>);
                                        ^
1 warning generated.
```

**Comment by davidbolvansky** [comment]
Not correct way to silence it via void.


`static_assert(!constant_invocable<[] { [i = int_limits.min()]() mutable { --i; (void)i; }(); }>);`

**Comment by JohelEGP** [comment]
That works. But I shouldn't need to do that, since `i` is indeed used, so the warning is bogus. My guess is that I mindlessly inserted a cast to `void` to silence the warning without noticing that `i` is indeed used. Now it doesn't work, hence the issue with the slipped useless cast.

**Comment by davidbolvansky** [comment]
i is set by - - i by then is not used. Seems ok to me.

**Comment by davidbolvansky** [comment]
.. also consider i-= 1; 

i is just set and then not used.

**Comment by JohelEGP** [comment]
That makes sense. I crafted my previous post with "unused" part in mind, missing the "but set" part of the warning. However, a reference to `i` is returned (for fundamental types) and then casted to `void`. I'm supposing that the warning was recently added. If there was consensus in its design that the case reported should issue a warning (perhaps because arbitrary expressions with program-defined types can't be supported like fundamental types), then this can be closed.

**Comment by AaronBallman** [comment]
> -Wunused-but-set-variable was added some months ago, just a few days ago it was extended to handle also pre/post inc/dec operators.
> 
> Imagine C++ code:
> 
> ```
> void test(int p = 42) {
>   p += 1;
> }
> ```
> 
> Clang prints: warning: parameter 'p' set but not used [-Wunused-but-set-parameter] void test(int p) { ^
> 
> I think is suboptimal and sooo confusing :( .. but this is how it was implemented in https://reviews.llvm.org/D100581

What do you find confusing with it? It's pointing to the parameter that's set but is not used.

> I would expect probably: warning: parameter 'p' set but not used [-Wunused-but-set-parameter] p += 1; ^ Looking at it, I dont see an easy way how to improve it.

Errr, I would be confused if a warning about parameters which are set but unused pointed to something other than the declaration of the parameter.

FWIW, I think the original example is a bug -- the cast to `void` should silence the diagnostic. However, while playing around, I noticed the behavior of this check is not what I would expect regarding template instantiations: https://godbolt.org/z/fz7sWqY9G

**Comment by JohelEGP** [comment]
Maybe a "set here" would help, like some other warnings also point to other relevant places, like some warnings involving template instantiations. Like this one here: https://godbolt.org/z/o5bz94ccq
```
mod.cpp:12:17: error: functions that differ only in their return type cannot be overloaded
  friend Target actual_downcast_guide(typename T::downcast_base);
         ~~~~~~ ^
mod.cpp:17:19: note: in instantiation of template class 'downcast_child<z, x>' requested here
export struct z : downcast_child<z, x> { };
                  ^
mod.cpp:12:17: note: previous declaration is here
  friend Target actual_downcast_guide(typename T::downcast_base);
         ~~~~~~ ^
```

**Comment by AaronBallman** [comment]
I think a note to identify where it was set is good idea, but there's two different behaviors to pick from: we could note every place the variable is set or we could note the first place we notice the variable is set. Noting every place the variable is set seems like the more useful behavior -- it shows you all the places you need to remove an assignment to silence the diagnostic, but removing the unused variable also tells you those same places. Only noting the first place is less chatty and easier to implement and works well when the correct way to silence the diagnostic is to use the variable instead of remove it.
